tool-openssl/req.cc

```diff
@@ -0,0 +1,666 @@
+        value = "";
+      } else if (len == 0 && attr.default_value) {
+        // Use default value
+        value = attr.default_value;
+      } else if (len > 0) {
+        // Use provided input
+        value = buffer;
+      } else {
+        // Empty input and no default - use empty string
+        value = "";
+      }
+
+      // Only add non-empty attributes
+      if (value && value[0]) {
+        X509_ATTRIBUTE* x509_attr = X509_ATTRIBUTE_create_by_NID(nullptr,
+                                                               attr.nid,
+                                                               MBSTRING_ASC,
+                                                               reinterpret_cast<const unsigned char*>(value),
+                                                               -1);
+        if (!x509_attr) {
+          fprintf(stderr, "Error creating attribute %s\n", attr.field_name);
+          return NULL;
+        }
+
+        if (!X509_REQ_add1_attr(req, x509_attr)) {
+          fprintf(stderr, "Error adding attribute %s to request\n", attr.field_name);
+          X509_ATTRIBUTE_free(x509_attr);
+          return NULL;
+        }
+
+        X509_ATTRIBUTE_free(x509_attr);
+      }
+    }
+  }
+
+  return subj;
+}
+
+static int make_certificate_request(X509_REQ *req, EVP_PKEY *pkey, std::string &subject_name, bool isCSR) {
+  X509_NAME *name;
+
+  // Set version
+  if (!X509_REQ_set_version(req, 0L))  // version 1
+    return 0;
+
+  if (subject_name.empty()) {// Prompt the user
+    name = prompt_for_subject(req, isCSR);
+  } else { // Parse user provided string
+    name = parse_subject_name(subject_name);
+    if (!name) {
+      return 0;
+    }
+  }
+
+  if (!X509_REQ_set_subject_name(req, name)) {
+    X509_NAME_free(name);
+    return 0;
+  }
+  X509_NAME_free(name);
+
+  if (!X509_REQ_set_pubkey(req, pkey)) {
+    return 0;
+  }
+
+  return 1;
+}
+
+static int req_password_callback(char *buf, int size, int rwflag, void *userdata) {
+  const char *prompt = "Enter PEM pass phrase:";
+  char verify_buf[1024];
+  int len;
+
+  // Display prompt
+  fprintf(stderr, "%s", prompt);
+  fflush(stderr);
+
+  // Get password
+  if (fgets(buf, size, stdin) == NULL) {
+    fprintf(stderr, "Error reading password\n");
+    return 0;
+  }
+
+  // Remove trailing newline
+  len = strlen(buf);
+  if (len > 0 && buf[len-1] == '\n')
+    buf[--len] = '\0';
+
+  // For encryption only (which is the case for req tool)
+  if (rwflag) {
+    // Verify password
+    fprintf(stderr, "Verifying - %s", prompt);
+    fflush(stderr);
+
+    if (fgets(verify_buf, sizeof(verify_buf), stdin) == NULL) {
+      fprintf(stderr, "Error reading verification password\n");
+      return 0;
+    }
+
+    // Remove trailing newline
+    int verify_len = strlen(verify_buf);
+    if (verify_len > 0 && verify_buf[verify_len-1] == '\n')
+      verify_buf[--verify_len] = '\0';
+
+    // Check if passwords match
+    if (strcmp(buf, verify_buf) != 0) {
```

Comment:
Unless one of the parameters is a literal, instead use `strncmp` to limit how far into the buffer you read:
```
if (strcmp(buf, verify_buf, BUF_SIZE) != 0) {
```